packages/test-renderer/src/createTestInstance.ts

```diff
@@ -12,15 +12,15 @@ export class ReactThreeTestInstance<TInstance extends Object3D = Object3D> {
   }
 
   public get instance(): Object3D {
-    return (this._fiber as unknown) as TInstance
+    return this._fiber as unknown as TInstance
   }
 
   public get type(): string {
     return this._fiber.type
   }
 
   public get props(): Obj {
-    return this._fiber.__r3f.memoizedProps
+    return this._fiber.__r3f?.memoizedProps ?? this._fiber
```

Comment:
Looks good. I'd do the same for the parent getter below.

Reply:
I don't think this should be returning the node if memoized props is not there, you're confusing props with the instance properties.

Reply:
done!

Reply:
What would you suggest then @joshuaellis ?

Reply:
Well you're not passing props to the object cause it's not rendered by react. So it should either return an empty object or not be accessible in the tree which would be even harder to do. I don't mind the former.

Reply:
Isn't the ReactThreeTestInstance.props() method just for retrieving a threejs component's props?

Reply:
the props passed to the component ie:

```js
<meshStandardMaterial color={0x00ff00} />
```

running `ReactThreeTestInstance.props` should return an object:

```js
{
  color: 0x00ff00
}
```

not all the properties of the THREE instance (which would include things like `alphaMap`.

Reply:
Ah all right! Understood. But those props are then a subset of the underlying Three-component, right? So returning just the component in case it was not mounted by r3f would return all props of this component. I think it can be useful in exactly this case – when the component exists in the scene but was not mounted by r3f. Would you agree?

Reply:
Hm I don't think I do agree. Then the function cross over with the '.instance' property of the node you're using.

I think this should be explicit where it returns props you provided to the component when rendering and if you didn't render it declaratively then it returns an empty object because there are no props.

Reply:
This functionality would be extremely useful to us! Finding everything with `.findByProps()` whether mounted by r3f or not saves so much hazzle. Returning an empty Object would in effect mean that the test-renderer seem to doesn't know about components not mounted via r3f although they are in the `scene`. In my opinion it enhances the `.findByProps()` method with a very useful and from the perspective of the client-code logical functionality, although, as you said, in this particular scenario the return-value of `.props` is equal to `.instance` for "manually" mounted components.

My recommendation is to add this change if there is no severe logic-breaking edge case I am overseeing.

Reply:
> you cannot use the .findByProps(), .findAll() etc. methods at all anymore because they throw. 

This isn't necessarily true:

```js
expect(() => scene.find((node) => node.instance.name === MANUALLY_MOUNTED_COMPONENT_NAME)).not.toThrow()
```

will not throw currently.

> as you said, in this particular scenario the return-value of .props is equal to .instance for "manually" mounted components.

Which makes the public facing API confusing. To someone new to an existing codebase with implemented tests, they might start to ask "why can I access some props but not all of them on certain components". It should act differently for two different nodes.

I'm not opposed to extending the API, we could have a new methods called `findByInstanceProperties` which could look like this:

```ts
  public findByInstanceProperties = (properties: Obj): ReactThreeTestInstance =>
    expectOne(this.findAllByInstanceProperties(properties), `with instance properties: ${JSON.stringify(properties)}`)

  public findAllByInstanceProperties = (properties: Obj): ReactThreeTestInstance[] =>
    findAll(this, (node: ReactThreeTestInstance) => Boolean(node.instance && matchProps(node.instance, properties)))
```

This would create a clear mental differentiation between the two APIs whilst including the feature you're requesting (which I have no problem with adding, it's just about how we add it).

What do you think? Also for more opinions cc @CodyJasonBennett

Reply:
The argument you made regarding the inconsistent API is true: returning a subset when the component is mounted via r3f and returning all props when it isn't is not good design. Your idea of adding a `.findByInstanceProps()` method is great! 
But then you'd write `.findByProps({name: "BOX-A"})` when you only have a "BOX-A" which is mounted via r3f and `.findByInstanceProps({name: "BOX-B"})` when you have "BOX-B" which isn't mounted via r3f. This doesn't seem quite elegant from the client-code perspective to me because you'd have to use different testing methods based on implementation details, which you actually want to ignore in black box tests that focus on user interaction. What do you say? Maybe there is a third even better way?

Reply:
Besides: I just tested your approach with the `.find()` method and it throws either because it uses `.getChildren()` which acesses `fiber.__r3f.objects`.

```
Cannot read properties of undefined (reading 'objects')
```

Reply:
I just noticed that `.findByProps()` is actually not user oriented as it is recommended in react-testing-library since it tests for things which the user cannot see (e.g. the `name` prop). It is comparable to `document.querySelector('[name=test])` and not to `renderer.getByText()` in react-testing-lib.

Reply:
I'd prefer to discuss this in an issue. There's a clear separation between react and three, and I don't want design to hold back fixes if we can avoid it.